### PR TITLE
Cache JSON of immutable ActivityPub representations

### PR DESCRIPTION
**Pull request by Gargron**
cc @ThibG

### CODE REVIEW

File: app/controllers/accounts_controller.rb [hunk 1]
```diff
@@ -2,7 +2,8 @@
 
 class AccountsController < ApplicationController
   include AccountControllerConcern
-  include SignatureVerification
```
Comment (Gargron) [inline app/controllers/accounts_controller.rb]: (This wasn't used)

File: app/controllers/accounts_controller.rb [hunk 2]
```diff
@@ -26,10 +27,11 @@ def show
...
       format.json do
-        render json: @account,
-               serializer: ActivityPub::ActorSerializer,
-               adapter: ActivityPub::Adapter,
-               content_type: 'application/activity+json'
+        skip_session!
+
+        render_cached_json(['activitypub', 'actor', @account.cache_key], content_type: 'application/activity+json') do
+          ActiveModelSerializers::SerializableResource.new(@account, serializer: ActivityPub::ActorSerializer, adapter: ActivityPub::Adapter)
+        end
```
Comment (ThibG) [inline app/controllers/accounts_controller.rb]: The rest seems fine, but isn't this going to cause Mastodon to serve stale account info (display name, avatar, header image, etc.)?

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/controllers/accounts_controller.rb (hunk 2 above)
Comment (Gargron) [reply app/controllers/accounts_controller.rb]: @ThibG: No. `#cache_key` method returns a combination of model name, ID and updated_at

Comment (ThibG) [reply app/controllers/accounts_controller.rb]: Alright, it's fine with me, then!

**Review by ClearlyClaire — APPROVED** [review]

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -123,11 +123,23 @@ def respond_with_error(code)
...
   def render_cached_json(cache_key, **options)
+    options[:expires_in] ||= 3.minutes
+    cache_key              = cache_key.join(':') if cache_key.is_a?(Enumerable)
+    content_type           = options.delete(:content_type) || 'application/json'
+
     data = Rails.cache.fetch(cache_key, { raw: true }.merge(options)) do
       yield.to_json
     end
 
     expires_in options[:expires_in], public: true
```
Comment (ThibG) [inline app/controllers/application_controller.rb]: Sorry I didn't catch that earlier, but you're unconditionally marking it as public here, which may cause problems.
Indeed, I believe it is possible to get private toots rendered with “public: true” using this code. This shouldn't immediately cause issues as “Set-Cookies” headers would be set, preventing the response to be cached, but it's still worrying.

**Pull request merged**